app/src/backend/ActivityService.ts

```diff
@@ -21,9 +21,241 @@ import type { Transaction } from "sequelize";
+        );
+
+      await inventoryValue.update({
+        id: inventoryId as string,
+        ...inventoryValueParams,
+        sectorId,
+        subSectorId,
+        subCategoryId,
+      });
+    }
+
+    return await inventoryValue.reload();
+  }
+
+  private static async updateGasValues({
+    gasValues,
+    activityValue,
+    transaction,
+    gases,
+  }: {
+    gasValues: GasValueInput[];
+    activityValue: ActivityValue;
+    transaction: Transaction;
+    gases: Gas[];
+  }) {
+    for (const gasValue of gasValues) {
+      let emissionsFactor: EmissionsFactor | null = null;
+      if (
+        gasValue.emissionsFactorId == null &&
+        gasValue.emissionsFactor != null
+      ) {
+        emissionsFactor = await db.models.EmissionsFactor.findOne({
+          where: { inventoryId: activityValue.inventoryValueId },
+          include: [
+            {
+              model: db.models.ActivityValue,
+              as: "activityValue",
+              where: { id: activityValue.id },
+            },
+          ],
+          transaction,
+        });
+
+        // don't edit emissions factors from other inventories or pre-defined ones (without inventoryId)
+        if (
+          emissionsFactor &&
+          emissionsFactor.inventoryId === activityValue.inventoryValueId
+        ) {
+          Object.assign(emissionsFactor, gasValue.emissionsFactor);
+          await emissionsFactor.save({ transaction });
+        } else {
+          emissionsFactor = await db.models.EmissionsFactor.create(
+            {
+              ...gasValue.emissionsFactor,
+              id: randomUUID(),
+              inventoryId: activityValue.inventoryValueId,
+            },
+            { transaction },
+          );
+        }
+      }
+
+      if (gasValue.emissionsFactor && !emissionsFactor) {
+        throw new createHttpError.InternalServerError(
+          "Failed to create an emissions factor",
+        );
+      }
+
+      delete gasValue.emissionsFactor;
+
+      if (gasValue.gasAmount == null) {
+        gasValue.gasAmount =
+          gases.find((gas) => gas.gas === gasValue.gas)?.amount ?? 0n;
+      }
+
+      await db.models.GasValue.upsert(
+        {
+          emissionsFactorId: emissionsFactor?.id ?? undefined,
+          ...gasValue,
+          id: gasValue.id ?? randomUUID(),
+          activityValueId: activityValue.id,
+          inventoryValueId: activityValue?.inventoryValueId,
+        },
+        { transaction },
+      );
+    }
+  }
+
+  public static async updateActivity({
+    id,
+    inventoryValueId,
+    inventoryValueParams,
+    activityValueParams,
+    gasValues,
+    dataSourceParams,
+  }: {
+    id: string;
+    activityValueParams: Omit<ActivityValueAttributes, "id">;
+    inventoryValueId: string | undefined;
+    inventoryValueParams: Omit<InventoryValueAttributes, "id"> | undefined;
+    gasValues: GasValueInput[] | undefined;
+    dataSourceParams: Omit<DataSourceAttributes, "datasourceId"> | undefined;
```

Comment:
I think generally you don't want to omit the IDs here but you might have found ways around that by using the relations...

Reply:
For the activityValue ID, I prefer to use the ID from the url, and my reasoning for using the relations for the inventoryValueId was to cater for situation where the user doesn't pass the inventoryValueParams as part of the patch body